aiida/tools/importexport/dbimport/utils.py

```diff
@@ -218,6 +218,8 @@ def deserialize_attributes(attributes_data, conversion_data):
 
 def deserialize_field(key, value, fields_info, import_unique_ids_mappings, foreign_ids_reverse_mappings):
     """Deserialize field using deserialize attributes"""
+    if key in ('attributes', 'extras'):
```

Comment:
was this broken before?
is there not test for this?

Reply:
So before, the attributes and extras fields were appended to the data after this deserialization. This was really just due to the implementation detail of the current archive format, whereby the node attributes and extras are "special cased" and stored in separate sections of the `data.json`.

In `ReaderJsonZip.iter_entity_fields`, you will see that I re-merge these fields into the fields dict, before yielding it.

I guess rather than having this test here, ideally we would change `aiida/tools/importexport/common/config.py::get_all_fields_info` to include these fields in the returned dict.
This would though necessitate a migration, to update this dict where it is stored in the `0.9`  archive (in `metadata.json`).

One thing I should mention here actually, is that really I want to eventually do a "second round" of refactoring of `import_data_dj`/`import_data_sqla`, to find a way for the process to actually make use of the `iter_entity_fields` iterator, and not have to read all the entities and their fields into memory at the same time.
This would not affect the current archive format, because you have to read the whole of `data.json` into memory anyway. But obviously with a new format this does not necessarily have to be the case.

Reply:
> I guess rather than having this test here, ideally we would change `aiida/tools/importexport/common/config.py::get_all_fields_info` to include these fields in the returned dict.
> This would though necessitate a migration, to update this dict where it is stored in the `0.9` archive (in the `all_fields_info` key of `metadata.json`).

Ok. I will anyhow need a migration for the group extras, so before merging this, I could give you write access to my fork to add this

> One thing I should mention here actually, is that really I want to eventually do a "second round" of refactoring of `import_data_dj`/`import_data_sqla`, to find a way for the process to actually make use of the `iter_entity_fields` iterator, and not have to read all the entities and their fields into memory at the same time.
> This would not affect the current archive format, because you have to read the whole of `data.json` into memory anyway. But obviously with a new format this does not necessarily have to be the case.

Right. I agree it's good to hold this off for a later round/PR (unless it just "falls out" naturally).

Reply:
> Ok. I will anyhow need a migration for the group extras, so before merging this, I could give you write access to my fork to add this

yep makes sense to do it in a single migration. I guess we want to give these fields a new `converter_type`, that basically tells `deserialize_attributes` to do nothing to the value.
Something like:

```python
def deserialize_attributes(attributes_data, conversion_data):
    """Deserialize attributes"""
    import datetime
    import pytz

   if conversion_data == "null":
       return attributes_data
```

Reply:
> Right. I agree it's good to hold this off for a later round/PR 

yep indeed 👍

FYI, this will also include probably passing `ArchiveData.node_data` to the writer as a query or generator for node extras/attributes (as opposed to the current iterator), so that it can perform a chunked `query.iterall(batch_size=x)` write into the archive and potentially (with a new format) not have to read all that data into memory at the same time.
(which I reckon is the largest source of data from the database).

Reply:
> Ok. I will anyhow need a migration for the group extras,

Just wanted to mention that Casper pointed out that, since the group extras are a simple addition (and we don't offer back-migrations), a migration is actually not needed to support group extras, i.e. we could also defer this migration to a later point.